### ISSUE TITLE
msi_ms7d25: Cannot get platform specific key to verify signatures. : (2)

### ISSUE THREAD

**Issue by deeplow** [Dasharo/meta-dts#77]
Dasharo Tool Suite fails on the MSI Z690 platform due to a wrong PATH. This probably affects other platforms.

When running an update it shows the following:
```
 msi_ms7d25: Cannot get platform specific key to verify signatures. : (2)
```

**Comment by pietrushnic** [comment]
@deeplow thank you for the report.

**Comment by Firminator** [comment]
I'm also affected by this. Just replying here for documentation purposes. I gather that the soon-to-be-released DTS v1.2.15 has the signature verification path already fixed. Wow! If only MSI, ASUS, Dell, HP, Toshiba or any of the other big wigs would act that fast fixing BIOS issue. They rather spend time on flooding the market with new models and dropping support fast :)

Reported in the Dasharo Matrix chat:

Hi there, question regarding DTS v1.2.14 ; after subscribing to the "1 year Dasharo Entry Subscription for Desktop" I received an email with Logs key, Download Key and Password. I'm trying to update the coreboot UEFI from Dasharo v1.1.1 for the MSI Pro Z690-A (WIFI) to v1.1.2 (which is only available for subscribers and supporters). I used option `4) Load DES keys` and double and triple-checked there is no typo, but it fails saying

> Something may be wrong with credentials. Please use option 4 to change DES keys and make sure that there is no typo.

'SecureBoot' and 'BIOS boot medium lock' and 'Enable SMM BIOS write protection' are all disabled in the coreboot UEFI .

When using option `5) Update Dasharo firmware` I'm being presented with a different error message:

> Cannot get platform specific key to verify signature. : (2)

Thanks.

**Comment by TomaszAIR** [comment]
@deeplow @Firminator problem with downloading GPG key for MSI releases is fixed in DTS rel v1.2.15 which is now available from iPXE by booting boot.dasharo.com/dts/dts.ipxe. Binaries to flash DTS image on USB are also available and can be found here https://github.com/Dasharo/meta-dts/releases/tag/v1.2.15. We apologise for the inconvenience.

**Comment by TomaszAIR** [comment]
@Firminator about the problem with keys

> Something may be wrong with credentials. Please use option 4 to change DES keys and make sure that there is no typo.

Could you please retest that on v1.2.15 release? Are you using MSI DDR4 version?

**Comment by Firminator** [comment]
Yes DDR4.

I just iPXE-booted DTS v1.2.15 from the Dasharo boot menu and used option 4 to load the DTS keys and it still says:

`Something may be wrong with credentials. Please use option 4 to change DES keys and make sure that there is no typo.`

**Comment by Firminator** [comment]
However when using option `5) Update Dasharo firmware` the previous error message

> Cannot get platform specific key to verify signature. : (2)

is now gone. So that seems to be fixed. :thumbsup:

**Comment by Firminator** [comment]
> Yes DDR4.
> 
> I just iPXE-booted DTS v1.2.15 from the Dasharo boot menu and used option 4 to load the DTS keys and it still says:
> 
> > Something may be wrong with credentials. Please use option 4 to change DES keys and make sure that there is no typo.

I think this here is the code: https://github.com/Dasharo/meta-dts/blame/main/meta-dts-distro/recipes-dts/dts/dts/dts#L165

**Comment by pietrushnic** [comment]
@Firminator I think it should go like this:
1. Please reply to the email where you received your credentials
2. We will check your credentials
3. If needed, we will regenerate credentials or fix our code

To be honest, you are the first person to report this kind of behavior, and we have a number of subscribers. I kind a fill there is something wrong with your credentials.

**Comment by Firminator** [comment]
OK thanks. Will do.  For what it's worth it does have a troublesome special character in the password... a character not easy to digest or pass along if not properly escaped...  a character usually excluded on a lot of online services:
`>`
(the infamous Greater-than character)

**Comment by pietrushnic** [comment]
Yeah, that can also be the problem. I will ask to switch to passphrases instead of those random character passwords.

**Comment by Firminator** [comment]
I send the email to your address. Thanks.

**Comment by pietrushnic** [comment]
@Firminator got it, thank you.

**Comment by TomaszAIR** [comment]
@Firminator today we tested your credentials, results below (I replaced keys with X,Y,Z).

```
  DTS version v1.2.16

  1) Dasharo HCL report - dump hardware information from this device
  3) Restore firmware from Dasharo HCL report
  4) Load DES keys
  5) Update Dasharo firmware
  8) Stop SSH server
  9) Shell
  10) Power off system
  11) Reboot system

Enter an option: 4


Enter logs key:                X

Enter firmware download key:   Y

Enter password:                Z
Dasharo DES credentials has been saved
Veryfing Dasharo DES credentials...
Waiting for network connection ...
Checking if board is Dasharo compatible.
Waiting for network connection ...
Dasharo DES credentials are correct and ready to use

  Logs key:         X
  Download key:     Y
  Password:         Z
  DTS version v1.2.16

  1) Dasharo HCL report - dump hardware information from this device
  3) Restore firmware from Dasharo HCL report
  4) Edit DES keys
  5) Update Dasharo firmware
  8) Stop SSH server
  9) Shell
  10) Power off system
  11) Reboot system

Enter an option: 9

Entering shell, to leave type exit and press Enter or press LCtrl+D

bash-5.1# dmidecode -s system-manufacturer
Micro-Star International Co., Ltd.
bash-5.1# dmidecode -s system-product-name
MS-7D25
bash-5.1# dmidecode -s baseboard-product-name
PRO Z690-A WIFI DDR4(MS-7D25)
```

At the end I run couple dmidecode commands just to confirm that we tests on the same platform.

How you use DTS in your case? USB or via iPXE?

If USB maybe you could mount it while booted to other OS and create file under path `/etc/cloud-pass` with your credentials on the USB? (so it would be `/path/to/mount/etc/cloud-pass`) If DTS boots and see that file it should already set needed variables and you should be able to update to v1.1.2.

If iPXE maybe you could start SSH server in DTS (option `8`), connect to it from different machine, start scripts with `dts` command then use option `4` to load keys and copy&paste them using that different machine? This is what I did while testing on MSI.

**Comment by Firminator** [comment]
Low and behold... SSH works using the same keys that don't work when running DTS directly on the device:
```
root@DasharoToolsSuite:~# dts
     ____             __                        ______            __        _____       _ __        
    / __ \____ ______/ /_  ____ __________     /_  __/___  ____  / /____   / ___/__  __(_) /____    
   / / / / __ `/ ___/ __ \/ __ `/ ___/ __ \     / / / __ \/ __ \/ / ___/   \__ \/ / / / / __/ _ \   
  / /_/ / /_/ (__  ) / / / /_/ / /  / /_/ /    / / / /_/ / /_/ / (__  )   ___/ / /_/ / / /_/  __/   
 /_____/\__,_/____/_/ /_/\__,_/_/   \____/    /_/  \____/\____/_/____/   /____/\__,_/_/\__/\___/   

  DTS version v1.2.16
 
  1) Dasharo HCL report - dump hardware information from this device``
  3) Restore firmware from Dasharo HCL report
  4) Load DES keys
  5) Update Dasharo firmware
  8) Stop SSH server
  9) Shell
  10) Power off system
  11) Reboot system
 
Enter an option: 4
 
Enter logs key:                my log key    
 
Enter firmware download key:   my download key
 
Enter password:                my password
Dasharo DES credentials has been saved
Veryfing Dasharo DES credentials...
Waiting for network connection ...
Checking if board is Dasharo compatible.
Waiting for network connection ...
Dasharo DES credentials are correct and ready to use

   Logs key:         my logs key
   Download key:     my download key
   Password:         my password
   DTS version v1.2.16
 
   1) Dasharo HCL report - dump hardware information from this device
   3) Restore firmware from Dasharo HCL report
   4) Edit DES keys
   5) Update Dasharo firmware
   8) Stop SSH server
   9) Shell
   10) Power off system
   11) Reboot system
 
Enter an option:
```

Since SSH from a remote box works I would think the bug is in how DTS script parses local user input?!

I'm going to try the USB method and see if pre-populating the keys into the credentials file @ /etc/cloud-pass

Thanks Thomasz for SSH workaround :+1:

**Comment by pietrushnic** [comment]
@Firminator I would suspect keyboard, locales, or something related. Please note no one else has this problem with input, and many people are trying. I'm pretty sure someone has your special characters, too, but the critical question is the keyboard and how the system receives codes from it. BTW is it maybe PS/2?

**Comment by Firminator** [comment]
Good thoughts. I'm using a USB keyboard and have PS/2 disabled in the coreboot UEFI. Locale is US but I haven't seen anywhere where it asks me which locale I wanna use, so I'm using whatever the DTS default locale is.
On a second thought though there is still a chance that I **did** fatfingered the 3 random passwords even after double and triple checking :shrug: I continue my tests.

**Comment by Firminator** [comment]
I created the /etc/cloud-pass file but what exactly do I have to paste into file? I can't find documentation for it.

Update: according to the code from dts starting at line 49 to 55 (https://github.com/Dasharo/meta-dts/blob/main/meta-dts-distro/recipes-dts/dts/dts/dts) all I have to do is paste the 3 keys each in a new line without any variables prefixing them since the code just parses/seds line by line using 1p, 2p, 3p :)

```  # Each time the main menu is displayed, check for DES credentials and export
  # them, if file exists
  if [ -e "${SE_credential_file}" ]; then
      export CLOUDSEND_LOGS_URL=$(sed -n '1p' < ${SE_credential_file} | tr -d '\n')
      export CLOUDSEND_DOWNLOAD_URL=$(sed -n '2p' < ${SE_credential_file} | tr -d '\n')
      export CLOUDSEND_PASSWORD=$(sed -n '3p' < ${SE_credential_file} | tr -d '\n')
      export DES_IS_LOGGED="true"
```
This works and now populates my keys automatically into DTS! Awesome.

```
    ____             __                        ______            __        _____       _ __
   / __ \____ ______/ /_  ____ __________     /_  __/___  ____  / /____   / ___/__  __(_) /____
  / / / / __ `/ ___/ __ \/ __ `/ ___/ __ \     / / / __ \/ __ \/ / ___/   \__ \/ / / / / __/ _ \
 / /_/ / /_/ (__  ) / / / /_/ / /  / /_/ /    / / / /_/ / /_/ / (__  )   ___/ / /_/ / / /_/  __/
/_____/\__,_/____/_/ /_/\__,_/_/   \____/    /_/  \____/\____/_/____/   /____/\__,_/_/\__/\___/


  Logs key:         my logs key
  Download key:     my download key
  Password:         my password
  DTS version v1.2.16

  1) Dasharo HCL report - dump hardware information from this device
  3) Restore firmware from Dasharo HCL report
  4) Edit DES keys
  5) Update Dasharo firmware
  8) Stop SSH server
  9) Shell
  10) Power off system
  11) Reboot system

Enter an option: 5

Waiting for network connection ...
Checking if board is Dasharo compatible.
Gathering flash chip and chipset information...
Flash information: vendor="Programmer" name="Opaque flash chip"
Flash size: 32M
Getting platform specific GPG key... Done
Checking for the latest Dasharo update available...
Current Dasharo version: 1.1.1
Latest available Dasharo version: 1.1.2
Downloading Dasharo firmware...Done

Please verify detected hardware!

Board vendor: Micro-Star International Co., Ltd.
System model: MS-7D25
Board model: PRO Z690-A WIFI DDR4(MS-7D25)

Does it match your actual specification? (Y|n) 
```

I'm gonna go ahead and press Y :100:

**Comment by Firminator** [comment]
This can be closed as the error 'Cannot get platform specific key to verify signatures' is fixed in DTS v1.2.16

**Comment by artur-rs** [comment]
@Firminator Thank you for your valuable feedback and time in this matter.
